Flaky test: ConnectorMockTest.testConnectorResourceMetricsScaledToZero

### Bug Description

This test seems to fail randomly when running the full unit and integration test suites. Instead, it always works when running in isolation.

```sh
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR] io.strimzi.operator.cluster.operator.assembly.ConnectorMockTest.testConnectorResourceMetricsScaledToZero(VertxTestContext)
[ERROR]   Run 1: ConnectorMockTest.testConnectorResourceMetricsScaledToZero(VertxTestContext) java.lang.AssertionError: 
Expected: is <0.0>
     but: was <1.0>
[ERROR]   Run 2: ConnectorMockTest.testConnectorResourceMetricsScaledToZero(VertxTestContext) java.lang.AssertionError: 
Expected: is <0.0>
     but: was <1.0>
[ERROR]   Run 3: ConnectorMockTest.testConnectorResourceMetricsScaledToZero(VertxTestContext) java.lang.AssertionError: 
Expected: is <0.0>
     but: was <1.0>
[ERROR]   Run 4: ConnectorMockTest.testConnectorResourceMetricsScaledToZero(VertxTestContext) java.lang.AssertionError: 
Expected: is <0.0>
     but: was <1.0>
[ERROR]   Run 5: ConnectorMockTest.testConnectorResourceMetricsScaledToZero(VertxTestContext) java.lang.AssertionError: 
Expected: is <0.0>
     but: was <1.0>
[ERROR]   Run 6: ConnectorMockTest.testConnectorResourceMetricsScaledToZero(VertxTestContext) java.lang.AssertionError: 
Expected: is <0.0>
     but: was <1.0>
[INFO] 
[INFO] 
[ERROR] Tests run: 2138, Failures: 1, Errors: 0, Skipped: 1
[INFO] 
```

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Strimzi version

main

### Kubernetes version

1.28

### Installation method

_No response_

### Infrastructure

_No response_

### Configuration files and logs

_No response_

### Additional context

_No response_

--- a/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java
+++ b/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java
@@ -142,6 +142,7 @@ public class KafkaConnectAssemblyOperator extends AbstractConnectOperator<Kubern
         Map<String, String> podAnnotations = new HashMap<>();
         Map<String, String> controllerAnnotations = new HashMap<>();
 
+        final boolean hasZeroReplicas = connect.getReplicas() == 0;
         final AtomicReference<String> image = new AtomicReference<>();
         final AtomicReference<String> desiredLogging = new AtomicReference<>();
         final AtomicReference<Deployment> deployment = new AtomicReference<>();
@@ -198,11 +199,11 @@ public class KafkaConnectAssemblyOperator extends AbstractConnectOperator<Kubern
                         )
                         .migrateFromDeploymentToStrimziPodSets(deployment.get(), podSet.get()))
                 .compose(i -> reconcilePodSet(reconciliation, connect, podAnnotations, controllerAnnotations, image.get()))
-                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus, desiredLogging.get(), connect.defaultLogConfig()))
+                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus, hasZeroReplicas, desiredLogging.get(), connect.defaultLogConfig()))
                 .onComplete(reconciliationResult -> {
                     StatusUtils.setStatusConditionAndObservedGeneration(kafkaConnect, kafkaConnectStatus, reconciliationResult.cause());
 
-                    if (connect.getReplicas() != 0) {
+                    if (!hasZeroReplicas) {
                         kafkaConnectStatus.setUrl(KafkaConnectResources.url(connect.getCluster(), namespace, KafkaConnectCluster.REST_API_PORT));
                     }
 
@@ -347,9 +348,10 @@ public class KafkaConnectAssemblyOperator extends AbstractConnectOperator<Kubern
      * @param connect The connector
      * @param connectStatus Status of the KafkaConnect  resource (will be used to set the available
      *                      connector plugins)
+     * @param scaledToZero  Indicated whether the related Connect cluster is currently scaled to 0 replicas
      * @return A future, failed if any of the connectors' statuses could not be updated.
      */
-    private Future<Void> reconcileConnectors(Reconciliation reconciliation, KafkaConnect connect, KafkaConnectStatus connectStatus, String desiredLogging, OrderedProperties defaultLogging) {
+    private Future<Void> reconcileConnectors(Reconciliation reconciliation, KafkaConnect connect, KafkaConnectStatus connectStatus, boolean scaledToZero, String desiredLogging, OrderedProperties defaultLogging) {
         String connectName = connect.getMetadata().getName();
         String namespace = connect.getMetadata().getNamespace();
         String host = KafkaConnectResources.qualifiedServiceName(connectName, namespace);
@@ -358,7 +360,7 @@ public class KafkaConnectAssemblyOperator extends AbstractConnectOperator<Kubern
             return Future.succeededFuture();
         }
 
-        if (connect.getSpec().getReplicas() == 0)   {
+        if (scaledToZero)   {
             return connectorOperator.listAsync(namespace, new LabelSelectorBuilder().addToMatchLabels(Labels.STRIMZI_CLUSTER_LABEL, connectName).build())
                 .compose(connectors -> Future.join(
                     connectors.stream().map(connector -> Annotations.isReconciliationPausedWithAnnotation(connector)

--- a/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java
+++ b/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java
@@ -142,7 +142,6 @@ public class KafkaConnectAssemblyOperator extends AbstractConnectOperator<Kubern
         Map<String, String> podAnnotations = new HashMap<>();
         Map<String, String> controllerAnnotations = new HashMap<>();
 
-        boolean hasZeroReplicas = connect.getReplicas() == 0;
         final AtomicReference<String> image = new AtomicReference<>();
         final AtomicReference<String> desiredLogging = new AtomicReference<>();
         final AtomicReference<Deployment> deployment = new AtomicReference<>();
@@ -199,11 +198,11 @@ public class KafkaConnectAssemblyOperator extends AbstractConnectOperator<Kubern
                         )
                         .migrateFromDeploymentToStrimziPodSets(deployment.get(), podSet.get()))
                 .compose(i -> reconcilePodSet(reconciliation, connect, podAnnotations, controllerAnnotations, image.get()))
-                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus, hasZeroReplicas, desiredLogging.get(), connect.defaultLogConfig()))
+                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus, desiredLogging.get(), connect.defaultLogConfig()))
                 .onComplete(reconciliationResult -> {
                     StatusUtils.setStatusConditionAndObservedGeneration(kafkaConnect, kafkaConnectStatus, reconciliationResult.cause());
 
-                    if (!hasZeroReplicas) {
+                    if (connect.getReplicas() != 0) {
                         kafkaConnectStatus.setUrl(KafkaConnectResources.url(connect.getCluster(), namespace, KafkaConnectCluster.REST_API_PORT));
                     }
 
@@ -348,10 +347,9 @@ public class KafkaConnectAssemblyOperator extends AbstractConnectOperator<Kubern
      * @param connect The connector
      * @param connectStatus Status of the KafkaConnect  resource (will be used to set the available
      *                      connector plugins)
-     * @param scaledToZero  Indicated whether the related Connect cluster is currently scaled to 0 replicas
      * @return A future, failed if any of the connectors' statuses could not be updated.
      */
-    private Future<Void> reconcileConnectors(Reconciliation reconciliation, KafkaConnect connect, KafkaConnectStatus connectStatus, boolean scaledToZero, String desiredLogging, OrderedProperties defaultLogging) {
+    private Future<Void> reconcileConnectors(Reconciliation reconciliation, KafkaConnect connect, KafkaConnectStatus connectStatus, String desiredLogging, OrderedProperties defaultLogging) {
         String connectName = connect.getMetadata().getName();
         String namespace = connect.getMetadata().getNamespace();
         String host = KafkaConnectResources.qualifiedServiceName(connectName, namespace);
@@ -360,13 +358,14 @@ public class KafkaConnectAssemblyOperator extends AbstractConnectOperator<Kubern
             return Future.succeededFuture();
         }
 
-        if (scaledToZero)   {
+        if (connect.getSpec().getReplicas() == 0)   {
             return connectorOperator.listAsync(namespace, new LabelSelectorBuilder().addToMatchLabels(Labels.STRIMZI_CLUSTER_LABEL, connectName).build())
-                    .compose(connectors -> Future.join(
-                            connectors.stream().map(connector -> maybeUpdateConnectorStatus(reconciliation, connector, null, zeroReplicas(namespace, connectName)))
-                                    .collect(Collectors.toList())
-                    ))
-                    .map((Void) null);
+                .compose(connectors -> Future.join(
+                    connectors.stream().map(connector -> Annotations.isReconciliationPausedWithAnnotation(connector)
+                            ? maybeUpdateConnectorStatus(reconciliation, connector, null, null)
+                            : maybeUpdateConnectorStatus(reconciliation, connector, null, zeroReplicas(namespace, connectName)))
+                        .collect(Collectors.toList())
+                )).map((Void) null);
         }
 
         KafkaConnectApi apiClient = connectClientProvider.apply(vertx);
@@ -463,7 +462,7 @@ public class KafkaConnectAssemblyOperator extends AbstractConnectOperator<Kubern
                                     if (!Util.matchesSelector(selector(), connect)) {
                                         LOGGER.debugCr(reconciliation, "{} {} in namespace {} was {}, but Connect cluster {} does not match label selector {} and will be ignored", connectorKind, connectorName, namespace, action, connectName, selector());
                                         return Future.succeededFuture();
-                                    } else if (connect.getSpec() != null && connect.getSpec().getReplicas() == 0) {
+                                    } else if (connect.getSpec() != null && connect.getSpec().getReplicas() == 0 && !Annotations.isReconciliationPausedWithAnnotation(resource)) {
                                         LOGGER.infoCr(reconciliation, "{} {} in namespace {} was {}, but Connect cluster {} has 0 replicas", connectorKind, connectorName, namespace, action, connectName);
                                         updateStatus(reconciliation, zeroReplicas(namespace, connectName), resource, connectorOperator);
                                         return Future.succeededFuture();

--- a/cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ConnectorMockTest.java
+++ b/cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ConnectorMockTest.java
@@ -15,6 +15,7 @@ import io.fabric8.kubernetes.client.dsl.base.PatchType;
 import io.micrometer.core.instrument.Gauge;
 import io.micrometer.core.instrument.MeterRegistry;
 import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.search.RequiredSearch;
 import io.netty.channel.ConnectTimeoutException;
 import io.strimzi.api.kafka.Crds;
 import io.strimzi.api.kafka.model.common.Condition;
@@ -57,6 +58,7 @@ import io.vertx.junit5.VertxExtension;
 import io.vertx.junit5.VertxTestContext;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.hamcrest.Matcher;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeAll;
@@ -447,11 +449,11 @@ public class ConnectorMockTest {
         waitForStatus(resource, connectorName, ConnectorMockTest.<KafkaConnector>statusIsForCurrentGeneration().and(ready()));
     }
 
-    public void waitForConnectorPaused(String connectName) {
+    public void waitForConnectorPaused(String connectorName) {
         Resource<KafkaConnector> resource = Crds.kafkaConnectorOperation(client)
                 .inNamespace(namespace)
-                .withName(connectName);
-        waitForStatus(resource, connectName, paused());
+                .withName(connectorName);
+        waitForStatus(resource, connectorName, paused());
     }
 
     public void waitForConnectorState(String connectorName, String state) {
@@ -2067,47 +2069,46 @@ public class ConnectorMockTest {
         String connectorName = "connector";
 
         KafkaConnect kafkaConnect = new KafkaConnectBuilder()
-                .withNewMetadata()
-                    .withNamespace(namespace)
-                    .withName(connectName)
-                    .addToAnnotations(Annotations.STRIMZI_IO_USE_CONNECTOR_RESOURCES, "true")
-                .endMetadata()
-                .withNewSpec()
-                    .withReplicas(0)
-                    .withBootstrapServers("my-kafka:9092")
-                .endSpec()
-                .build();
+            .withNewMetadata()
+                .withNamespace(namespace)
+                .withName(connectName)
+                .addToAnnotations(Annotations.STRIMZI_IO_USE_CONNECTOR_RESOURCES, "true")
+            .endMetadata()
+            .withNewSpec()
+                .withReplicas(0)
+                .withBootstrapServers("my-kafka:9092")
+            .endSpec()
+            .build();
 
         Crds.kafkaConnectOperation(client).inNamespace(namespace).resource(kafkaConnect).create();
         waitForConnectReady(connectName);
 
         KafkaConnector connector = defaultKafkaConnectorBuilder()
-                .editMetadata()
-                    .withName(connectorName)
-                    .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, connectName)
-                    .addToAnnotations(Annotations.ANNO_STRIMZI_IO_PAUSE_RECONCILIATION, "true")
-                .endMetadata()
-                .build();
+            .editMetadata()
+                .withName(connectorName)
+                .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, connectName)
+                .addToAnnotations(Annotations.ANNO_STRIMZI_IO_PAUSE_RECONCILIATION, "true")
+            .endMetadata()
+            .build();
 
         Crds.kafkaConnectorOperation(client).inNamespace(namespace).resource(connector).create();
-        waitForConnectorNotReady(connectorName, "RuntimeException", "Kafka Connect cluster 'cluster' in namespace " + namespace + " has 0 replicas.");
-
-        MeterRegistry meterRegistry = metricsProvider.meterRegistry();
-        Tags tags = Tags.of("kind", KafkaConnector.RESOURCE_KIND, "namespace", namespace);
+        waitForConnectorPaused(connectorName);
 
         Promise<Void> reconciled = Promise.promise();
         kafkaConnectOperator.reconcileAll("test", namespace, ignored -> reconciled.complete());
 
         Checkpoint async = context.checkpoint();
         reconciled.future().onComplete(context.succeeding(v -> context.verify(() -> {
-            Gauge resources = meterRegistry.get(MetricsHolder.METRICS_RESOURCES).tags(tags).gauge();
-            assertThat(resources.value(), is(1.0));
-
-            kafkaConnectOperator.metrics().pausedConnectorsResourceCounter(namespace); // to create metric, otherwise MeterNotFoundException will be thrown
-            Gauge resourcesPaused = meterRegistry.get(MetricsHolder.METRICS_RESOURCES_PAUSED).tags(tags).gauge();
-            assertThat(resourcesPaused.value(), is(0.0));
+            String[] tags = new String[]{"kind", KafkaConnector.RESOURCE_KIND, "namespace", namespace};
+            assertGaugeMetricMatches(MetricsHolder.METRICS_RESOURCES, tags, is(1.0));
+            assertGaugeMetricMatches(MetricsHolder.METRICS_RESOURCES_PAUSED, tags, is(1.0));
             async.flag();
         })));
+    }
+
+    private void assertGaugeMetricMatches(String name, String[] tags, Matcher<Double> matcher) {
+        RequiredSearch requiredSearch = metricsProvider.meterRegistry().get(name).tags(tags);
+        assertThat(requiredSearch.gauge().value(), matcher);
     }
 
     @Test

